src/proto.rs

```diff
@@ -1285,6 +1287,22 @@ impl Decoder for WebsocketProtocol {
 
                 mask::frame(masking_key, to_unmask, self.payload_in & 3);
             }
+
+            if fin && opcode == OpCode::Text {
+                let utf8_valid_up_to = self.utf8_valid_up_to.unwrap_or_default();
+                let (should_fail, _) = utf8::should_fail_fast(
+                    // SAFETY: offset + utf8_valid_up_to is the index until which utf8 was
+                    // validated for this frame and therefore guaranteed to be in bounds.
+                    unsafe {
+                        src.get_unchecked(offset + utf8_valid_up_to..offset + payload_length)
+                    },
+                    true,
+                );
+
+                if should_fail {
+                    return Err(Error::Protocol(ProtocolError::InvalidUtf8));
+                }
+            }
```

Comment:
This should just use `utf8::parse_str` and propagate the error. `parse_str` uses `simdutf8::basic`, which is faster than `should_fail_fast`'s usage of `simdutf8::compat`.